api/provider/aws/builds.go

```diff
@@ -128,64 +243,242 @@ func (p *AWSProvider) BuildDelete(app, id string) (*structs.Build, error) {
+		ConsistentRead: aws.Bool(true),
+		Key: map[string]*dynamodb.AttributeValue{
+			"id": &dynamodb.AttributeValue{S: aws.String(id)},
+		},
+		TableName: aws.String(buildsTable(a.Name)),
+	}
 
+	res, err := p.dynamodb().GetItem(req)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	// failed builds could have an empty manifest
-	if len(entries) == 0 {
-		return nil
+	if res.Item == nil {
+		return nil, fmt.Errorf("no such build: %s", id)
 	}
 
-	urls := []string{}
+	build := p.buildFromItem(res.Item, a.Outputs["Settings"])
 
-	for name, _ := range entries {
-		img := fmt.Sprintf("%s/%s-%s:%s", os.Getenv("REGISTRY_HOST"), a.Name, name, b.Id)
+	return build, nil
+}
 
-		if registryId := a.Outputs["RegistryId"]; registryId != "" {
-			img = fmt.Sprintf("%s.dkr.ecr.%s.amazonaws.com/%s:%s.%s", registryId, os.Getenv("AWS_REGION"), a.Outputs["RegistryRepository"], name, b.Id)
-		}
+func (p *AWSProvider) BuildList(app string) (structs.Builds, error) {
+	a, err := p.AppGet(app)
+	if err != nil {
+		return nil, err
+	}
 
-		urls = append(urls, img)
+	req := &dynamodb.QueryInput{
+		KeyConditions: map[string]*dynamodb.Condition{
+			"app": &dynamodb.Condition{
+				AttributeValueList: []*dynamodb.AttributeValue{&dynamodb.AttributeValue{S: aws.String(a.Name)}},
+				ComparisonOperator: aws.String("EQ"),
+			},
+		},
+		IndexName:        aws.String("app.created"),
+		Limit:            aws.Int64(20),
+		ScanIndexForward: aws.Bool(false),
+		TableName:        aws.String(buildsTable(a.Name)),
 	}
 
-	imageIds := []*ecr.ImageIdentifier{}
-	registryId := ""
-	repositoryName := ""
+	res, err := p.dynamodb().Query(req)
+	if err != nil {
+		return nil, err
+	}
 
-	for _, url := range urls {
-		if match := regexpECR.FindStringSubmatch(url); match != nil {
-			registryId = match[1]
-			repositoryName = match[3]
+	builds := make(structs.Builds, len(res.Items))
 
-			imageIds = append(imageIds, &ecr.ImageIdentifier{
-				ImageTag: aws.String(match[4]),
-			})
-		} else {
-			return errors.New("URL not valid ECR")
+	for i, item := range res.Items {
+		builds[i] = *p.buildFromItem(item, a.Outputs["Settings"])
+	}
+
+	return builds, nil
+}
+
+func (p *AWSProvider) BuildRelease(b *structs.Build) (*structs.Release, error) {
+	releases, err := p.ReleaseList(b.App)
+	if err != nil {
+		return nil, err
+	}
+
+	r := structs.NewRelease(b.App)
+	newId := r.Id
+
+	if len(releases) > 0 {
+		r = &releases[0]
+	}
+
+	r.Id = newId
+	r.Created = time.Time{}
+	r.Build = b.Id
+	r.Manifest = b.Manifest
+
+	a, err := p.AppGet(b.App)
+	if err != nil {
+		return r, err
+	}
+
+	err = p.ReleaseSave(r, a.Outputs["Settings"], a.Parameters["Key"])
+	if err != nil {
+		return r, err
+	}
+
+	b.Release = r.Id
+	err = p.BuildSave(b, "")
+	return r, err
+}
+
+// BuildSave creates or updates a build item in DynamoDB. It takes an optional
+// bucket argument, which if set indicates to PUT Log data into S3
+func (p *AWSProvider) BuildSave(b *structs.Build, bucket string) error {
```

Comment:
I understand that. If you look at the interface definition its a more agnostic definition:

```
func BuildSave(b *structs.Build, logdir string) error
```

The reason being that this bucket is not available in the environment, it's something different for every app.

But on AWS since build has app.Name I can remove this from the definition and do the app and settings bucket lookup in the BuildSave method.